crates/extrinsics/src/lib.rs

```diff
@@ -347,29 +384,58 @@ pub async fn fetch_contract_info(
     contract: &AccountId32,
     rpc: &LegacyRpcMethods<DefaultConfig>,
     client: &Client,
-) -> Result<Option<ContractInfo>> {
-    let info_contract_call = api::storage().contracts().contract_info_of(contract);
-
+) -> Result<ContractInfo> {
     let best_block = get_best_block(rpc).await?;
 
-    let contract_info_of = client
+    let contract_info_address = dynamic(
+        "Contracts",
+        "ContractInfoOf",
+        vec![Value::from_bytes(contract)],
+    );
+    let contract_info = client
         .storage()
         .at(best_block)
-        .fetch(&info_contract_call)
-        .await?;
+        .fetch(&contract_info_address)
+        .await?
+        .ok_or_else(|| {
+            anyhow!(
+                "No contract information was found for account id {}",
+                contract
+            )
+        })?;
+    #[derive(scale_decode::DecodeAsType, Debug)]
+    pub struct BoundedVec<T>(pub ::std::vec::Vec<T>);
+    #[derive(scale_decode::DecodeAsType, Debug)]
+    struct ContractInfoOf {
+        trie_id: BoundedVec<u8>,
+        code_hash: CodeHash,
+        storage_items: u32,
+        storage_item_deposit: Balance,
+    }
+    #[derive(scale_decode::DecodeAsType)]
+    struct DepositAccount {
+        deposit_account: AccountId32,
+    }
 
-    match contract_info_of {
-        Some(info_result) => {
-            let convert_trie_id = hex::encode(info_result.trie_id.0);
-            Ok(Some(ContractInfo {
-                trie_id: convert_trie_id,
-                code_hash: info_result.code_hash,
-                storage_items: info_result.storage_items,
-                storage_item_deposit: info_result.storage_item_deposit,
-            }))
+    let total_balance: Balance = match contract_info.as_type::<DepositAccount>() {
+        Ok(account) => {
+            // StorageVersion >= 10 and < 15 contains deposit in separate account
```

Comment:
This comment could be expanded to explain a bit what is going on here, I took me a little bit to figure it out. I understand it as: newer versions of `pallet_contracts` have an additional field `deposit_account`, so we first attempt to decode that and get the balance for that account. If that fails we assume that it is the older version without that field. Then I am assuming that on line 431 this decoding will succeed for both cases because the `contract_info.as_type::<DepositAccount>()` will have already consumed the additional field if present?

Anyway it would be a nice present for future maintainers since there are no tests for this code.

Reply:
Maybe we could add a unit test for this part of the logic, if we extract a pure function e.g. `fn get_deposit_account_id(contract_info: DecodedValueThunk) -> Result<AccountId32>`

Reply:
> This comment could be expanded to explain a bit what is going on here, I took me a little bit to figure it out. I understand it as: newer versions of `pallet_contracts` have an additional field `deposit_account`, so we first attempt to decode that and get the balance for that account. If that fails we assume that it is the older version without that field. Then I am assuming that on line 431 this decoding will succeed for both cases because the `contract_info.as_type::<DepositAccount>()` will have already consumed the additional field if present?
> 
> Anyway it would be a nice present for future maintainers since there are no tests for this code.

Versions [>=10, <15] have additional field `deposit_account` in `ContractInfoOf` struct. If it exists the contract's deposit is there, otherwise contract's deposit is on the main contract's account.

Regarding decoding on the line 431: It succeeds for both case, because we decode only common part of data.

Reply:
> DecodedValueThunk

I would like to do it, but I do not see the public API to create DecodedValueThunk object

Reply:
Hm I guess you have to construct it with https://github.com/paritytech/subxt/blob/43809ee5928a77eed5438f7d61f24fe0183f970a/subxt/src/dynamic.rs#L44. Anyway if it is too much trouble then no need to bother, but would be nice for this to be tested.

Reply:
I will add it, thanks